x-pack/filebeat/input/httpjson/input.go

```diff
@@ -0,0 +1,320 @@
+	workerWg     sync.WaitGroup     // Waits on worker goroutine.
+}
+
+type requestInfo struct {
+	URL        string
+	ContentMap common.MapStr
+}
+
+// NewInput creates a new httpjson input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+	// Build outlet for events.
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap input.Context's Done channel with a context.Context. This goroutine
+	// stops with the parent closes the Done channel.
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputContext.Done:
+		case <-inputCtx.Done():
+		}
+	}()
+
+	// If the input ever needs to be made restartable, then context would need
+	// to be recreated with each restart.
+	workerCtx, workerCancel := context.WithCancel(inputCtx)
+
+	in := &httpjsonInput{
+		config: conf,
+		log: logp.NewLogger("httpjson").With(
+			"url", conf.URL),
+		outlet:       out,
+		inputCtx:     inputCtx,
+		workerCtx:    workerCtx,
+		workerCancel: workerCancel,
+	}
+
+	in.log.Info("Initialized httpjson input.")
+	return in, nil
+}
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *httpjsonInput) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("httpjson input worker has started.")
+			defer in.log.Info("httpjson input worker has stopped.")
+			defer in.workerWg.Done()
+			defer in.workerCancel()
+			if err := in.run(); err != nil {
+				in.log.Error(err)
+				return
+			}
+		}()
+	})
+}
+
+// createHTTPRequest creates an HTTP/HTTPs request for the input
+func (in *httpjsonInput) createHTTPRequest(ctx context.Context, ri *requestInfo) (*http.Request, error) {
+	b, _ := json.Marshal(ri.ContentMap)
+	body := bytes.NewReader(b)
+	req, err := http.NewRequest(in.config.HTTPMethod, ri.URL, body)
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(ctx)
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("User-Agent", userAgent())
+	if in.config.APIKey != "" {
+		req.Header.Set("Authorization", in.config.APIKey)
+	}
+	return req, nil
+}
+
+// processHTTPRequest processes HTTP request, and handles pagination if enabled
+func (in *httpjsonInput) processHTTPRequest(ctx context.Context, client *http.Client, req *http.Request, ri *requestInfo) error {
+	for {
+		msg, err := client.Do(req)
+		if err != nil {
+			return errors.New("failed to do http request. Stopping input worker - ")
```

Comment:
I'm curious about what the trailing ` - ` is for?

Reply:
Some text will be added at the end when printed, and this - make it reads better.